### ISSUE TITLE
EventStore Events has No Items to Display and Indicates an Error Message

### ISSUE THREAD

**Issue by brianstringfellow** [Microsoft/service-fabric#285]
A new service fabric cluster (v6.4.622.9590) was deployed in Azure using the page [here ](https://docs.microsoft.com/en-us/azure/service-fabric/service-fabric-diagnostics-eventstore) to add the relevant section to the ARM template. Deployment was successful and the cluster is healthy. The service `fabric:/System/EventStoreService` is ready and healthy.

Error message in UI:
![image](https://user-images.githubusercontent.com/22823524/51064893-db5c4800-15b6-11e9-9010-b1be20e8614c.png)

Error:  null failed.
Code: E_ACCESSDENIED
Message: Access Denied.

Everything else in the UI is accessible and works. What could be causing the error?

**Comment by ibabou** [comment]
@brianstringfellow, can you please try hitting the UI from a different fresh browser window and select the cert when prompted? and then let me know if the issue is still persistent?. Also, is this happening with all new clusters you tried to create or only with this instance?. EventStore service is served through the same gateway as other system services, so it is a bit weird to get E_ACCESSDENIED.

**Comment by brianstringfellow** [comment]
Hi @ibabou - the problems persists. I have tried both Chrome and Edge. The problem occurs with any new cluster based on the ARM template used for deployment. Does the EventStore use a particular port? We have our NSGs locked down based upon the published service fabric cluster needs and the needs of our applications.

**Comment by ibabou** [comment]
The https port that EventStore service will be listening on is dynamic. but, the request goes first through the gateway (19080). and then if all nodes can talk to each others within the range of ports specified for all applications, gateway on one node should be able to talk to the primary of EventStore service if it is running on a different one ?.

**Comment by brianstringfellow** [comment]
Are there any logs anywhere that I can view to perhaps get a hint at what may be wrong?
I've looked at the primary node's event logs and application insights. I have the WadCfg setup for the standard service fabric listeners.

**Comment by juho-hanhimaki** [comment]
I am observing the same behaviour on my cluster. 

Perhaps the issue is that i am using Azure AD to authenticate to cluster (instead of certificate). Does EventStore support Azure AD authentication?

![image](https://user-images.githubusercontent.com/12390182/51668933-00e04d00-1fcc-11e9-945d-f56635812994.png)

**Comment by juho-hanhimaki** [comment]
Still seeing this same issue on my cluster:

{
  "Error": {
    "Code": "E_ACCESSDENIED",
    "Message": "Access Denied."
  }
}

I am wondering if the root cause of this has something do with same error message i get with BackupRestoreService on the same cluster:
https://github.com/MicrosoftDocs/azure-docs/issues/23747

**Comment by ibabou** [comment]
Hi @juho-hanhimaki , Currently the service requires a X.509 security to be used for authentication. and yes it has been architected similar to the backup restore service by having its own http/https endpoint, so it requires a Cert for a secured cluster.

**Comment by juho-hanhimaki** [comment]
@ibabou 
Just to make sure we don't have misunderstanding here my cluster is secured with certificate.

![image](https://user-images.githubusercontent.com/12390182/52083946-8c984180-25a9-11e9-9415-69ee983da9ad.png)

I tried my cluster with pure X.509 authentication earlier today. I my test i just removed the "azureActiveDirectory" configuration node from my ARM template and deployed the change. Yet I think I encountered the exact same problem. But I would have to double check as honestly I've been trying many different things here in hope to solve the issues.

Either way I don't fully understand how the method of authentication would have effect because either way the cluster is secured with a certificate and even AAD secured cluster can be connected to using the cluster certificate if I am not mistaken.

But if you indeed are correct about the authentication method being the root cause it should be documented clearly because many of the customers will opt in to use AAD authentication and assume it should work just as well as certificate based authentication. Finding out stuff like this the hard way is waste of time for the customers.

**Comment by juho-hanhimaki** [comment]
Just spun up a new cluster but removed the AD authentication part. Still see the same error message when trying to view events.

**Comment by ibabou** [comment]
Hi @juho-hanhimaki, is there any specific NSG rules you've ? EventStore service uses a port from applications ports pool for its https server. and any gateway that receives the request should be able to talk to the primary of EventStore service to process it. If not, and this is just a regular new SFRP cluster, please open a ticket to customer support (who can help creating an incident) so we can look into the cluster configuration. 
Other couple of questions, do you see the EventStore service healthy ? Were you able to figure out the reason of not being able to reach BackupRestore service you mentioned above ?

Thanks!

**Comment by ibabou** [comment]
@brianstringfellow , so did you use a template similar to this when creating the cluster:
https://github.com/Azure-Samples/service-fabric-cluster-templates/blob/master/5-VM-Windows-1-NodeTypes-Secure-NSG/azuredeploy.json
Or did you add the NSG rules later ?

because as you can see in this sample template, the applications ports range is opened between the vms:
          {
            "name": "allowSvcFabApplication",
            "properties": {
              "access": "Allow",
              "destinationAddressPrefix": "*",
              "destinationPortRange": "[concat(variables('nt0applicationStartPort'), '-', variables('nt0applicationEndPort'))]",
              "direction": "Inbound",
              "priority": 3940,
              "protocol": "*",
              "sourceAddressPrefix": "*",
              "sourcePortRange": "*",
              "description": "allow fabric application ports within the vnet"
            }
          },

**Comment by juho-hanhimaki** [comment]
@ibabou I was finally able to figure out what is causing the issue. It seems the EventStore and BackupRestoreService both can't be reached when cluster certificate is bound by common name.

That is too bad because using common name to bind the certificate is considered a best practice:
https://docs.microsoft.com/en-us/azure/service-fabric/service-fabric-best-practices-security#secure-a-service-fabric-cluster-certificate-by-common-name
https://docs.microsoft.com/fi-fi/azure/service-fabric/service-fabric-production-readiness-checklist

Hopefully something can be done to work around this issue?

Here are the specific changes I made to fix the issue:

Before:

        "virtualMachineProfile": {
         ...
                    "certificate": {
                      "commonNames": [
                        "[parameters('certificateCommonName')]"
                      ],
                      "x509StoreName": "[variables('certificateStoreName')]"
                    }

      "type": "Microsoft.ServiceFabric/clusters",
       ...
        "certificateCommonNames": {
          "commonNames": [
            {
              "certificateCommonName": "[parameters('certificateCommonName')]",
              "certificateIssuerThumbprint": ""
            }
          ],
          "x509StoreName": "[variables('certificateStoreName')]"
        }

After:

        "virtualMachineProfile": {
         ...
                    "certificate": {
                      "thumbprint": "<mythumbrint>",
                      "x509StoreName": "[variables('certificateStoreName')]"
                    }

      "type": "Microsoft.ServiceFabric/clusters",
       ...
        "certificate": {
          "thumbprint": "<mythumbrint>",
          "x509StoreName": "[variables('certificateStoreName')]"
        },

**Comment by ibabou** [comment]
Hi @juho-hanhimaki , can you please share the resourceGroup/region/cluster where you deployed the cluster with the common names configuration ? and if you can, please share as well the deployment correlation id ?

Thanks,
Ibrahim

**Comment by brianstringfellow** [comment]
The scenario described by @juho-hanhimaki is exactly what I have encountered. I have an old cluster that uses a certificate thumbprint and I was able to get EventStore to work there. On a new cluster that uses the certificate common name, that's where I see this problem.

**Comment by juho-hanhimaki** [comment]
@ibabou I just emailed you the cluster details.

**Comment by hrushib** [comment]
As discussed offline on email thread, this is happening as `certificateIssuerThumbprint` is left empty. Please configure this with appropriate value and this will start accepting requests.

**Comment by brianstringfellow** [comment]
Hi @ibabou, is the issue really closed? As I understand from the following Microsoft page, the cluster services should work with common name and `certificateIssuerThumbprint` is blank. The same was pointed out by juho-hanhimaki above.

https://docs.microsoft.com/en-us/azure/service-fabric/service-fabric-cluster-change-cert-thumbprint-to-cn 

```
{
    "apiVersion": "2018-02-01",
    "type": "Microsoft.ServiceFabric/clusters",
    "name": "[parameters('clusterName')]",
    "location": "[parameters('clusterLocation')]",
    "dependsOn": [
        "[concat('Microsoft.Storage/storageAccounts/', variables('supportLogStorageAccountName'))]"
    ],
    "properties": {
        "addonFeatures": [
            "DnsService",
            "RepairManager"
        ],
        "certificateCommonNames": {
            "commonNames": [
                {
                    "certificateCommonName": "[parameters('certificateCommonName')]",
                    "certificateIssuerThumbprint": ""
                }
            ],
            "x509StoreName": "[parameters('certificateStoreValue')]"
        },
    ...
```

**Comment by juho-hanhimaki** [comment]
Yep, I am not fully happy with this either. There exists no good reason to require certificateIssuerThumbprint for a certificate that is totally valid and issued by trusted root CA.

Cluster also doesn't require it. It's only EventStore and BackupRestoreService that do.

Managing IssuerThumbprints is unnecessary burden as its totally normal that the Issuer might change when new certificates are renewed/new certificates are bought.

**Comment by aljo-microsoft** [comment]
@brianstringfellow  and @juho-hanhimaki and @MicahMcKittrick-MSFT

Please assign this to me as a doc bug.

I will fix our documentation to always reflects declaring an issuer thumbprint, as it is a best practice to do.

**Comment by juho-hanhimaki** [comment]
I can't believe you would recommend using issuer thumbprint when service fabric team has announced in blog post that thumbprints are basically evil. If other customers feel the same please, let your opinion be known.

https://blogs.msdn.microsoft.com/azureservicefabric/2018/04/27/secrets-management-through-certificate-common-names/

Please reconsider.

**Comment by dragav** [comment]
@brianstringfellow @juho-hanhimaki are your certificates trusted? Are your clusters working fine otherwise? Thanks.

**Comment by juho-hanhimaki** [comment]
@dragav Yes trusted (AddTrust External CA Root) and working fine except for the inability to connect to BackupRestoreService and EventStore.

**Comment by brianstringfellow** [comment]
@dragav Yes, trusted (Trusted Secure Certificate Authority 5) and working fine with services, aside from EventStore.

**Comment by juho-hanhimaki** [comment]
This is what @hrushib said in our email conversation regarding "certificateIssuerThumbprint" parameter:
> BRS uses this value explicitly to validate incoming certificate

So if that statement is accurate at least there should be a clear understanding of why the cluster works but BackupRestoreService (and EventStore) don't.

The empty certificateIssuerThumbprint has been discussed previously here:
https://github.com/MicrosoftDocs/azure-docs/issues/16440#issuecomment-428687793

**Comment by dragav** [comment]
@brianstringfellow @juho-hanhimaki thank you both for the prompt confirmation. We support specifying intermediate thumbprints as a means to strengthen the certificate validation, and also for situations where the root is not trusted (it's pinning the issuer vs pinning the leaf's thumbprint). As of the current release, issuer pinning is not required although recommended. We will follow up internally to clear up this discrepancy.

Thanks for your patience and support.

**Comment by aljo-microsoft** [comment]
@juho-hanhimaki 
I wrote our team blog posting you referenced related to common name. It is recommended that you pin your issuer thumbprint, and I will update our documentation on this shortly.

**Comment by cwe1ss** [comment]
Does this mean that we should populate the common name AND the thumbprint in the template?

**Comment by rdunne76** [comment]
Same issue here - actually migrated from thumbprint to Common Name as  per best practice and experience the same problem re thumbprint and cert based authentication.
Can we get an indication on when this will be resolved as it is impacting  on our ability to monitor environments

**Comment by dragav** [comment]
A fix has been checked in, and is making its way to the upcoming releases. I don't have an ETA for when this will roll out. For the time being, please pin the issuer thumbprints to work around this issue.

I'd like to reiterate that declaring issuer thumbprints is definitely safer than simply relying on a trusted root. All expected issuers of a given PKI should be declared (the field takes a list of comma-separated values) to prevent outages following a rollover.

**Comment by lanfeust69** [comment]
This may be related to Azure/service-fabric-issues#1553.

**Comment by BaMi83** [comment]
We just had this issue as well after changing the cert from thumbprint to common name, our cluster version was 8.0 CU1, issue still there, i contacted MS and they helped me, it is fixed now, the only thing you have to do is to restart the primary **EventStoreService** the primary will flip over to another node, reboot this one as well, do this till the primary role is back to a node you just rebooted. 
![image](https://user-images.githubusercontent.com/26976733/121997409-9dc50000-cdaa-11eb-8c52-3c457d166237.png)
you will see quickly reconfiguring and after this the events are all visible again!
![image](https://user-images.githubusercontent.com/26976733/121997471-b33a2a00-cdaa-11eb-9dd1-1d5e9f095f45.png)
